### DISCUSSION TITLE
Error setting userProvided embedder

### DISCUSSION THREAD

**Discussion by miguelisidoro** [meilisearch/product#749] [Feedback & Feature Proposal]
Hello,

We are trying to set our index embedder to a userProvided approach.
This is our request and response.

![image](https://github.com/meilisearch/product/assets/5429152/432b9da5-5102-40f4-b51b-5c4da0cb89e9)

However, we are getting the following error in the associated task;

![image](https://github.com/meilisearch/product/assets/5429152/2d66b9f3-f971-4367-a6b3-8e16eed9ae10)

Why are we getting this error? Do we have to delete all documents from the index before setting the embedder?

Thanks

**Comment by dureuill** [comment]
> Why are we getting this error? 

When adding a `userProvided` embedder, all existing documents need to have `_vectors.<embedder-name>` (here `_vectors.default`) defined to one or multiple embeddings (arrays of floats)

> Do we have to delete all documents from the index before setting the embedder?

That's an option. You can also update all documents to add the missing `_vectors.<embedder-name>` field before adding your embedder.

**Reply by miguelisidoro** [reply]
Hello @dureuill,

Thanks for the response.

We took a look at https://www.meilisearch.com/docs/learn/experimental/vector_search#generate-custom-embeddings-manually, and the _vectors field would have to be something like:

"_vectors": {"default": [0, 0.8, -0.2]}

Assuming that the embedder will be called "default". Can you confirm?

Thanks

**Reply by dureuill** [reply]
yes that's correct :+1:

**Reply by miguelisidoro** [reply]
Hello @dureuill, thanks for the confirmation.

**Reply by miguelisidoro** [reply]
Hello @dureuill,

I just rebuilt my search index and all items have a _vectors field with a vector of dimension 1536, the size of the vector returned by Azure Open AI Embedding API.

However, when setting the user provided embedder, I get the following error:

![image](https://github.com/meilisearch/product/assets/5429152/172d4eac-0849-4d0b-974a-dd280e71d40b)

The user provided embedder is being set like this:

![image](https://github.com/meilisearch/product/assets/5429152/0efd5afd-b390-45dd-b8c5-75fe70ba5a9b)

Example of a document with the _vectors filled.

![image](https://github.com/meilisearch/product/assets/5429152/63c4fec5-acef-4d0d-9697-bba6c369bb84)

What am I doing wrong here?

Any way to get the document in the index mentioned in the error and checking if it has a valid vector?

Any way to get all documents that have null in the _vectors field?

Thanks

**Reply by dureuill** [reply]
Hello

> What am I doing wrong here?

Looks like one or more of your documents doesn't have a correctly shaped `_vectors` field

> Any way to get the document in the index mentioned in the error and checking if it has a valid vector?

Run `GET http://<your-meilisearch-instance>/indexes/<your-index>/documents/Default_4488r fwm613nqy1ph9tw2jbvwd` to get the document by its id mentioned in the error.

> Any way to get all documents that have null in the _vectors field?

Not easily, I'm afraid. In v1.8, you *could* set `_vectors` as filterable, but this is strongly discouraged for performance reasons (it will index all the individual floats in the vectors) and won't be possible in v1.9 and higher (disabled due to the performance impact)

I'll be discussing this use case with the team because it is true that this is a desirable use case when adding a user provided embedder

**Reply by miguelisidoro** [reply]
Hello @dureuill,

Thanks for your response.
There are indeed some documents with _vectors = null.
I will solve this and test again.

Thanks
